### PR TITLE
Add `override_repo` and `inject_repo`

### LINKED ISSUE #23580: Can't use module_override to override a repo fetched inside module extension
### Description of the bug:

We have a module extension that uses a repo rule with rctx.download to download an artifact. We use it from the top level MODULE.bazel as:

```
fetcher = use_extension("//src/repo/bazel/rules:fetcher_extension.bzl", "fetcher")
fetcher.artifact_with_plat(
     name = "artifact1",
     url = y
     hash = z
)
use_repo("artifact1")
```

This works fine. The artifact ends up in my extensions folder as `+fetcher+artifact1`

How do I then override this artifact using override_module? `--override_module=artifact1=...` does not work, and `--override_module=+fetcher+artifact1=...` does not work.

### Which category does this issue belong to?

External Dependency

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

_No response_

### Which operating system are you running Bazel on?

windows

### What is the output of `bazel info release`?

release 7.3.1

### If `bazel info release` returns `development version` or `(@non-git)`, tell us how you built Bazel.

_No response_

### What's the output of `git remote get-url origin; git rev-parse HEAD` ?

_No response_

### If this is a regression, please try to identify the Bazel commit where the bug was introduced with bazelisk --bisect.

_No response_

### Have you found anything relevant by searching the web?

no

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by fmeum**
Work towards #19301
Fixes #23580

RELNOTES: `override_repo` and `inject_repo` can be used to override and inject repos in module extensions.

### CODE REVIEW

File: site/en/external/extension.md [hunk 1]
```diff
@@ -174,6 +174,13 @@ several repo visibility rules:
         the repo visible to the module instead of the extension-generated repo
         of the same name.
 
+### Overriding and adding repos
+
+The root module can use
+[`override_repo`](/rules/lib/globals/module#override_repo) to override or add
```
Comment (Wyverald) [inline site/en/external/extension.md:180]: curious why you mentioned "adding repos" here. What does it mean for a repo to be added to an extension? I can't think of a case where it meaningfully matters.

If you do have a use case in mind, it would be nice to add it to the doc here (would be nice to have an example for overriding too, actually)

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java [hunk 2]
```diff
@@ -784,15 +841,20 @@ public void include(String label, StarlarkThread thread)
             doc =
                 "A list of labels pointing to patch files to apply for this module. The patch files"
                     + " must exist in the source tree of the top level project. They are applied in"
-                    + " the list order.",
+                    + " the list order."
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:844]: these should be moved to the other PR

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java [hunk 3]
```diff
@@ -637,6 +644,56 @@ public void useRepo(
...
+      StarlarkThread thread)
+      throws EvalException {
+    ModuleThreadContext context = ModuleThreadContext.fromOrFail(thread, "override_repo()");
+    context.setNonModuleCalled();
+    if (context.shouldIgnoreDevDeps()) {
+      // Ignore calls early as they may refer to repos that are dev dependencies (or this is not the
+      // root module).
+      return;
+    }
+    ImmutableList<StarlarkThread.CallStackEntry> stack = thread.getCallStack();
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:687]: why do we need the call stack? everything in MODULE.bazel files should be at the "top" layer, so `thread.getCallerLocation()` should suffice, no?

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java [hunk 4]
```diff
@@ -121,12 +127,25 @@ List<ModuleExtensionUsageBuilder> getExtensionUsageBuilders() {
   }
 
   static class ModuleExtensionUsageBuilder {
+
+    private record RepoOverride(
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java:131]: can move this up one layer -- since it's being used one layer up anyhow

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java [hunk 5]
```diff
@@ -121,12 +127,25 @@ List<ModuleExtensionUsageBuilder> getExtensionUsageBuilders() {
...
+        String exportedName,
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java:132]: these names should probably be updated to match the semantics of `override_repo`. So `exportedName` should be `overriddenRepoName`, and `localRepoName` should be `overridingRepoName`. same elsewhere

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java [hunk 6]
```diff
@@ -171,6 +191,23 @@ void addImport(String localRepoName, String exportedName, String byWhat, Locatio
...
     }
 
+    public void addRepoOverride(
+        String exportedName,
+        String localRepoName,
+        ImmutableList<StarlarkThread.CallStackEntry> stack)
+        throws EvalException {
+      RepositoryName.validateUserProvidedRepoName(exportedName);
+      RepositoryName.validateUserProvidedRepoName(localRepoName);
+      if (repoOverrides.containsKey(exportedName)) {
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java:201]: optional -- this `if` can be merged with the follow `put` (see `addRepoNameUsage` above):

```java
RepoOverride collision = repoOverrides.put(...);
if (collision != null) { ... }
```

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java [hunk 7]
```diff
@@ -197,6 +234,26 @@ ModuleExtensionUsage buildUsage() throws EvalException {
...
         builder.setIsolationKey(Optional.empty());
       }
+
+      for (var override : repoOverrides.entrySet()) {
+        String exportedName = override.getKey();
+        String localRepoName = override.getValue().localRepoName;
+        if (!context.repoNameUsages.containsKey(localRepoName)) {
+          throw Starlark.errorf(
+                  "The repo exported as '%s' by module extension '%s' is overridden with '%s', but"
+                      + " no repo is imported under this name",
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java:244]: s/imported/visible ? technically it doesn't need to be a repo imported using `use_repo`.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java [hunk 8]
```diff
@@ -54,6 +55,11 @@ public class ModuleThreadContext extends StarlarkThreadContext {
   private final Map<String, ModuleOverride> overrides = new HashMap<>();
   private final Map<String, RepoNameUsage> repoNameUsages = new HashMap<>();
 
+  private final Map<String, ModuleExtensionUsageBuilder.RepoOverride> reposSubjectToRepoOverride =
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java:58]: these two can similarly (see other comment) be renamed to just `overriddenRepos` and `overridingRepos`.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java [hunk 9]
```diff
@@ -260,6 +317,43 @@ public InterimModule buildModule(@Nullable Registry registry) throws EvalExcepti
       }
       extensionUsages.add(extensionUsageBuilder.buildUsage());
     }
+    // Check for chains in repo overrides: If A is overridden with B, and B is overridden with C,
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java:320]: I think this could probably be more simply formulated as "a repo cannot both be overriding and overridden". Makes the error reporting a bit cleaner as well.

**Review by Wyverald — COMMENTED** [review]
Overall looks great! I haven't finished reviewing yet (mostly just looked at ModuleFileGlobals & ModuleThreadContext); will continue later.

File: site/en/external/extension.md (hunk 1 above)
Comment (fmeum) [reply site/en/external/extension.md]: I added examples and also split `override_repo` into `override_repo` and `inject_repo`.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java (hunk 3 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:689]: That wouldn't give us stack traces and thus source file snippets. Inconsistent Starlark error reporting is a pet peeve of mine. I think that consistently showing stack traces with source snippets is better than sometimes just showing locations, even if the stack traces only have one frame.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java (hunk 9 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java]: I updated the comment, but I'm not sure how to update the error reporting. I would still like to suggest shortening A -> B -> C to just A -> C and wouldn't know how to do that with the new shorter error message.

**Comment by fmeum** [comment]
As I still see many folks specify only a single repo per `use_repo`, I wonder whether we should use the plural form for `inject_repo` and `override_repo`.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java (hunk 9 above)
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java]: IMO we don't need to suggest the shortening, but I don't feel strongly about this. So up to you what to do here.

File: site/en/external/extension.md (hunk 1 above)
Comment (Wyverald) [reply site/en/external/extension.md]: these are great, thanks!

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java (hunk 3 above)
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:689]: for my curiosity, do you have a paste handy of the stack traces with source snippets? How does it work with multiple frames actually -- does each frame get a line shown, a la Python?

If we're using a (single-frame) stack here, we should ideally use it everywhere else, too. (by that I mean `use_repo` etc.) But also why do we store just a `Location` in `ModuleExtensionUsage.RepoOverride`? Maybe those should store stacks too.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java [hunk 10]
```diff
@@ -185,6 +207,10 @@ public Builder addTag(Tag value) {
       return this;
     }
 
+    @CanIgnoreReturnValue
+    public abstract Builder setRepoOverrides(
+        ImmutableMap<String, RepoOverride> extensionLocalToApparentName);
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java:212]: rename param

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java [hunk 11]
```diff
@@ -130,10 +131,26 @@ public final boolean getHasNonDevUseExtension() {
...
+   * <p>This is only non-empty for root module usages and repos that override other repos are not
+   * themselves overridden.
+   */
+  public abstract ImmutableMap<String, RepoOverride> getRepoOverrides();
+
   public abstract Builder toBuilder();
 
   public static Builder builder() {
-    return new AutoValue_ModuleExtensionUsage.Builder();
+    return new AutoValue_ModuleExtensionUsage.Builder().setRepoOverrides(ImmutableMap.of());
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java:153]: is this default value necessary? seems like we should make sure it's set in prod code, in exchange for a bit more verbose test code.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java [hunk 12]
```diff
@@ -130,10 +131,26 @@ public final boolean getHasNonDevUseExtension() {
...
+      // The apparent name of the overriding repo in the root module.
+      String overridingRepoName,
+      // Whether this override should apply to an existing repo.
+      boolean mustExist,
+      Location location) {}
+
+  /**
+   * Maps repo names local to the extension to repo they are overridden by.
+   *
+   * <p>This is only non-empty for root module usages and repos that override other repos are not
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java:145]: The "and repos that override ..." part confuses me. Why is it worth pointing out here?

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java [hunk 13]
```diff
@@ -130,10 +131,26 @@ public final boolean getHasNonDevUseExtension() {
     return getProxies().stream().anyMatch(p -> !p.isDevDependency());
   }
 
+  @GenerateTypeAdapter
+  public record RepoOverride(
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java:135]: need javadoc for this

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java [hunk 14]
```diff
@@ -130,10 +131,26 @@ public final boolean getHasNonDevUseExtension() {
...
+      // The apparent name of the overriding repo in the root module.
+      String overridingRepoName,
+      // Whether this override should apply to an existing repo.
+      boolean mustExist,
+      Location location) {}
+
+  /**
+   * Maps repo names local to the extension to repo they are overridden by.
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java:143]: for a bit more clarity: "Contains information about overrides that apply to repos generated by this extension. Keyed by the extension-local repo name." (The value type meaning is self evident.)

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/SingleExtensionUsagesValue.java [hunk 15]
```diff
@@ -56,13 +57,16 @@ public abstract class SingleExtensionUsagesValue implements SkyValue {
   /** The repo mappings to use for each module that used this extension. */
   public abstract ImmutableMap<ModuleKey, RepositoryMapping> getRepoMappings();
 
+  public abstract ImmutableMap<String, RepositoryName> getRepoOverrides();
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/SingleExtensionUsagesValue.java:60]: add javadoc -- particularly, what do the key and value types mean (this is less obvious than the same field in `MdouleExtensionUsage`)

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphValue.java [hunk 16]
```diff
@@ -103,27 +106,54 @@ public static BazelDepGraphValue createEmptyDepGraph() {
    */
   public abstract ImmutableMap<ModuleExtensionId, String> getExtensionUniqueNames();
 
+  /**
+   * For each module extension, a mapping from the name of the repo exported by the extension to the
+   * canonical repo name of the repo that should override it (if any).
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphValue.java:111]: super nit -- "canonical ~~repo~~ name of the repo"

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java [hunk 17]
```diff
@@ -198,6 +203,38 @@ private static String makeUniqueNameCandidate(ModuleExtensionId id, int attempt)
                 + extensionNameDisambiguator);
   }
 
+  private static ImmutableTable<ModuleExtensionId, String, RepositoryName> resolveRepoOverrides(
+      ImmutableMap<ModuleKey, Module> depGraph,
+      ImmutableTable<ModuleExtensionId, ModuleKey, ModuleExtensionUsage> extensionUsagesTable,
+      ImmutableMap<ModuleExtensionId, String> extensionUniqueNames,
+      ImmutableBiMap<RepositoryName, ModuleKey> canonicalRepoNameLookup) {
+    RepositoryMapping rootModuleMapping =
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java:211]: optional: name this `rootModuleMappingWithoutOverrides`. might be too long though, so I don't feel strongly.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java [hunk 18]
```diff
@@ -198,6 +203,38 @@ private static String makeUniqueNameCandidate(ModuleExtensionId id, int attempt)
...
+    ImmutableTable.Builder<ModuleExtensionId, String, RepositoryName> repoOverridesBuilder =
+        ImmutableTable.builder();
+    for (var extensionId : extensionUsagesTable.rowKeySet()) {
+      var rootUsage = extensionUsagesTable.row(extensionId).get(ModuleKey.ROOT);
+      if (rootUsage != null) {
+        for (var override : rootUsage.getRepoOverrides().entrySet()) {
+          repoOverridesBuilder.put(
+              extensionId,
+              override.getKey(),
+              rootModuleMapping.get(override.getValue().overridingRepoName()));
```
Comment (Wyverald) [inline src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java:231]: just realized that we can't actually override a repo with one using an arbitrary canonical repo name. Probably not a big problem, if one at all...

**Review by Wyverald — COMMENTED** [review]
haven't looked at tests yet, but this overall looks good to me. Thanks for getting this in place!

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java (hunk 3 above)
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:689]: > for my curiosity, do you have a paste handy of the stack traces with source snippets? How does it work with multiple frames actually -- does each frame get a line shown, a la Python?

Never mind this, an example is in the tests (and indeed it's like Python). The other question still stands.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java [hunk 19]
```diff
@@ -75,6 +79,18 @@ public ModuleThreadContext(
     this.includeLabelToCompiledModuleFile = includeLabelToCompiledModuleFile;
   }
 
+  record RepoOverride(
+      String overriddenRepoName,
+      String overridingRepoName,
+      boolean mustExist,
+      ModuleExtensionUsageBuilder extensionUsageBuilder,
```
[the remark this answers is not in the document]
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleThreadContext.java]: You are right, it's probably not worth it. I removed that logic and replaced the mutable field with just the extension name.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java (hunk 18 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelDepGraphFunction.java]: I would call that a feature as it would be a reason to hardcode these names. I think that it makes sense to have this only be possible via command line flags such as `--override_repository`, which have a more "temporary" nature.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java (hunk 3 above)
Comment (fmeum) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:692]: We just store a `Location` since we can't show two stacks for one error. The error will be on the `override_repo` call, but at least it references the import that makes the overriding repo overridden.

And yes, I agree that we should use it in more places, but I would like to do that in a follow-up.

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java (hunk 12 above)
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleExtensionUsage.java:148]: sorry, I'm still confused.

1. "This is only non-empty for root module usages" stands by itself, right? If so, maybe add a comma (or even a period) before "and repos that ...".
2. Why does the fact that overriding repos cannot be overridden mean "overrides do need to applied recursively"? What does applying an override recursively even mean?

File: src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java (hunk 3 above)
Comment (Wyverald) [reply src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java:692]: OK, that's fair.

File: src/test/java/com/google/devtools/build/lib/testutil/FoundationTestCase.java [hunk 20]
```diff
@@ -78,6 +83,18 @@ public final void initializeFileSystemAndDirectories() throws Exception {
     scratch.file(rootDirectory.getRelative("WORKSPACE").getPathString());
     scratch.file(rootDirectory.getRelative("MODULE.bazel").getPathString());
     root = Root.fromPath(rootDirectory);
+
+    // Let the Starlark interpreter know how to read source files.
+    EvalException.setSourceReaderSupplier(
```
Comment (Wyverald) [inline src/test/java/com/google/devtools/build/lib/testutil/FoundationTestCase.java:88]: just to double check, why is this part required for this PR only, not before? the `withCallStack` stuff was added in a previous commit, right? Is it because we never made any test assertions with the callstack before this PR?

Comment (fmeum) [reply src/test/java/com/google/devtools/build/lib/testutil/FoundationTestCase.java:88]: We had test assertions, but I missed that those didn't contain any source code snippets, just the `Location` information.

**Comment by fmeum** [comment]
> Why does the fact that overriding repos cannot be overridden mean "overrides do need to applied recursively"? What does applying an override recursively even mean?

I removed this comment, it's actually no longer relevant at that point since the overriding repo has already been resolved to a canonical name. I was referring to the "no chains" requirement relevant in `BazelDepGraphFunction#resolveRepoOverrides`, but you are totally right about this not mattering at the location you commented on.

**Review by Wyverald — APPROVED** [review]
🚀!!

**Comment by Wyverald** [comment]
Hmm. I have no idea why, but `command_profiler_test` seems to be consistently failing. Presubmit runs for other PRs don't seem to be affected. Baffling.

**Comment by fmeum** [comment]
> Hmm. I have no idea why, but `command_profiler_test` seems to be consistently failing. Presubmit runs for other PRs don't seem to be affected. Baffling.

That was my update to JDK 23 (oopsie...). I rebased onto master, which has the revert.

**Comment by fmeum** [comment]
@bazel-io fork 7.4.0
